### DISCUSSION TITLE
Question re end switch (reset) behavior

### DISCUSSION THREAD

**Discussion by hmmueller** [Paciente8159/uCNC#720] [Q&A] [answered]
Using uCNC 1.10.beta (master), running with PlatformIO's serial monitor, $21=1 (i.e. hard end switches enabled).
When I run into an end switch, all stops - fine. But then, I get two or three different behaviors - it seems at random:

* Sometimes, I have to reset (the Arduino) and then continue (with a $X).
* But most times, after doing the Arduino reset, I am immediately told that the end switch is triggered, and again a reset is asked for -> endless loop = I have to short the respective end switch so that uCNC will turn the motor.
* At times, after a full power down/up of Arduino (including removing the USB), uCNC seems to allow a $X.

Now, I don't plan to run into the end switches when all is up and running. But I'd like to understand what's going on - in the final machine, I wont be able to short end switches in case I run into one.

H.M.

**Comment by Paciente8159** [comment, marked as answer]
Depends on when you trigger the limit switch.

- If the machine was stopped it will allow you to continue with a simple unlock $X.
- If the machine was moving when the limit was activated it will require the user to do a soft-reset before allowing any other commands to the board.
- The power down/up is just a hard-reset. It does the same as a soft-reset, just a bit deeper.

This follows more or less the Grbl principles.

The idea is safety. If a limit was triggered while in motion it will not allow any other commands except a soft-reset to allow the user to inspect the machine for any safety issue and then requires it to acknowledge the machine is ok to be operated by sending a soft reset before unlocking it. This will force the user to restart the program it was running from start because there is a good possibility the  machine lost position from the abrupt stop.

You should be able to soft reset the machine and unlock it with a limit triggered and even run it. This assumes that the user recognized the problem and know what is doing. But as soon as that limit changes state it will lock again.

You might be getting micro state changes because a limit is at the edge of being opened/close and changes states for a brief period.
I recommend having a good noise isolation circuit or you will have lots of false triggers on your limits switches.

Also note that limits are active all the time while you are running. If you start a program that sends your machine outside of it's limits the switches will prevent a catastrophic crash.

**Reply by hmmueller** [reply]
That sounds all good. Only ... I think I'm not sure what "soft-reset" is? - is this connecting ESTOP to Vcc for a moment? (which I did not want to connect, see previous thread)

Re noise isolation: I did the "full program" - first, each end switch gets an RC ...

![Endschalter](https://github.com/user-attachments/assets/3c919d2d-8a02-4b8e-965e-d5c5bc822988)

.. and also, I separate spindle + coolant pump power energy from stepper and end switch wires; stepper motor wires and spindle power wire are shielded:

![SeparateEnergyChains](https://github.com/user-attachments/assets/0f6fd77a-32f9-4582-9c87-75238279ad0e)

I hope this is enough :-).

H.M.

**Reply by Paciente8159** [reply]
Soft reset is done when you send a Ctrl+X command to the board. It's almost similar to a Hard/Power reset. The only difference is that it does not run the initial configuration of the board and µCNC substystems again.

About control pins, limit pins etc... This is all configurable.
By default they have internal weak pull ups enabled. So if you forget to connect a switch, of a wire breaks...that input gets triggered.

So no...you don't need to connect ESTOP or FHOLD or any of those pins to 5V. You just need to force them to go to GND to disable them.

About the switch circuit. I actually prefer to use an opto coupler and run the limit switch side of the circuit at higher voltages (12V, 24V or whatever I'm using for the steppers). This makes more immune to noise IMHO.

![20240803_143413](https://github.com/user-attachments/assets/9f09a2f6-4c41-42ab-b30c-7dfbec23caee)

**Reply by hmmueller** [reply]
Yes, higher voltage is more immunity. Still, many people are happy with a simple capacitor of 100 nF or the like. I'll go with my 5V RC circuit - again, when it fails, I'll have learned something. As of now, it works (running the spindle with various speeds while stepping along), but those were only short tests, no real workload.

H.M.
